NiFi-4338: Add documents for how to use SSL protocol from HDFS processors

tasanuma opened the pull request:
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [x] Is your initial contribution a single, squashed commit?

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [x] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

tasanuma commented:
I did some system tests for this change and checked it works fine. But it is hard to write a unit test for using SWebHdfs from NiFi and I was not able to do.

This is my first contribution to NiFi. Please let me know if there is any problem.

bbende commented:
Hello, thank you for contributing!

Is there a way to pass the SSL/TLS values to Hadoop client without setting system properties?

Setting system properties will impact everything running in the JVM and could cause other running processors/libraries to have issues.

tasanuma commented:
Hello Bryan, thanks for the review!

I see. Other processors which use SSL Context Service use SSLContext class to pass the SSL/TLS values. I will investigate whether hadoop client can use it or not.

tasanuma commented:
I found that Hadoop has some properties for SSL protocols.
https://hadoop.apache.org/docs/r2.8.0/hadoop-mapreduce-client/hadoop-mapreduce-client-core/EncryptedShuffle.html

After setting the properties, I confirmed that SWebHdfs can be used from HDFS processors.

I would like to add the instruction to the documents of HDFS processors in this pull request.

Thanks!

tasanuma commented:
I reverted the last commit and added some documents for how to use SSL protocol as 'additionalDetails.html' in HDFS processors. Please review it when you are free.

ijokarumawak commented:
@tasanuma Thanks for writing the detailed documentations, those are very helpful! Are the contents of each additionalDetails.html the same among those processors?

If so, we can consolidate it to a single additionalDetails.html, such as PutHDFS (or whichever), then add a short description about swebhdfs in other XXXHDFS processors '@CapabilityDescription', like "To use swebhdfs, see PutHDFS 'additional details' documentation.". Together with '@SeeAlso', user can reach to the documentation when needed. It would provide better maintainability. How do you think?

tasanuma commented:
@ijokarumawak Thanks for the review and the comments! Yes, all the additionalDetails.html is the same. I agree with you that putting it in the document of PutHDFS. I updated my pull request based on your advice. Since I don't want Parquet processors depend on HDFS processors, I also added PutParquet/additionalDetails.html.

joewitt commented on nifi-nar-bundles/nifi-hadoop-bundle/nifi-hdfs-processors/src/main/java/org/apache/nifi/processors/hadoop/CreateHadoopSequenceFile.java:
```diff
@@ -64,7 +64,8 @@
 @Tags({"hadoop", "sequence file", "create", "sequencefile"})
-@CapabilityDescription("Creates Hadoop Sequence Files from incoming flow files")
+@CapabilityDescription("Creates Hadoop Sequence Files from incoming flow files."
+        + " If you want to use SSL-secured file system like swebhdfs, please see the 'SSL Configuration' topic of the 'Additional Details' of PutHDFS.")
```
I think we should avoid adding the 'if you want to use SSL-Secured file system" entry in all these descriptions.  If it is an option of the processor then on that option this comment can exist.

ijokarumawak replied:
@joewitt I agree, the explanation is not a 'capability description' rather a documentation for configuration detail.

Is there any better place to write a common topic like this for components within a NAR package? Something like 'package.html' in Javadoc, describing what the NAR includes and common configuration, what type of flow user can construct with provided processors ... etc. Is there any improvement going on in 'Extension Registry' sub project for such documentation? If not, does it sounds helpful? If so, I will raise a JIRA to add 'bundle.html' or 'nar.html' (Probably with Markdown support). How do you think?

tasanuma commented:
Thanks for the review and the comments, @joewitt and @ijokarumawak. I understand. I simply added the description to the option in the parent processor of HDFS.

I created a common additionalDetails.html in PutHDFS for now, but later it would be useful if there is a common space which is refered from processors.

joewitt commented:
thanks @tasanuma . I agree.  Am reviewing now and if checks out will merge shortly.

tasanuma commented:
Thanks for reviewing and committing it, @joewitt and @ijokarumawak!
